core/modules/user/user.module

```diff
@@ -1158,13 +1158,16 @@ function user_menu() {
     'type' => MENU_DEFAULT_LOCAL_TASK,
   );
 
-  $items['user/register'] = array(
-    'title' => 'Create new account',
-    'page callback' => 'backdrop_get_form',
-    'page arguments' => array('user_register_form'),
-    'access callback' => 'user_register_access',
-    'type' => MENU_LOCAL_TASK,
-  );
+  // Do not render the registration tab if the site is in maintenance mode.
+  if (!state_get('maintenance_mode', FALSE)) {
+    $items['user/register'] = array(
+      'title' => 'Create new account',
+      'page callback' => 'backdrop_get_form',
+      'page arguments' => array('user_register_form'),
+      'access callback' => 'user_register_access',
+      'type' => MENU_LOCAL_TASK,
+    );
+  }
```

Comment:
I think a smarter way of handling this may be to update `user_register_access()`, which could check maintenance mode. That would also the need to menu rebuild.